File: recurrent-learn/recurrent_learn.py
import sys

from pydantic.dataclasses import dataclass


@dataclass
class CorrectOrNot:
    first: bool
    second: bool


def problems_to_be_solved(correct_or_nots: list[CorrectOrNot]) -> list[int]:
    problems: list[int] = list()
    for i, correct_or_not in enumerate(correct_or_nots):
        if not correct_or_not.first or not correct_or_not.second:
            problems.append(i + 1)
    return problems


def parse_line(line: str) -> CorrectOrNot:
    line = line.strip()
    elems = line.split(" ")
    correct_or_not = CorrectOrNot(first=True, second=True)
    if elems[0] == "n":
        correct_or_not.first = False
    if elems[1] == "n":
        correct_or_not.second = False
    return correct_or_not


def run():
    lines = sys.stdin.readlines()
    correct_or_nots = list(map(parse_line, lines[1:]))
    print(correct_or_nots)
    problems = problems_to_be_solved(correct_or_nots)
    if len(problems) == 0:
        print(0)
    else:
        print(len(problems))
        for p in problems:
            print(p)


if __name__ == "__main__":
    run()

File: recurrent-learn/recurrentlearn.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

type CorrectOrNot struct {
	First  bool
	Second bool
}

type Solver interface {
	Solve(correctOrNots []CorrectOrNot) []int
}

type solverImpl struct{}

func new() Solver {
	return &solverImpl{}
}

func (s *solverImpl) Solve(correctOrNots []CorrectOrNot) []int {
	problems := []int{}
	for i, correctOrNot := range correctOrNots {
		if (!correctOrNot.First) || (!correctOrNot.Second) {
			problems = append(problems, i+1)
		}
	}
	return problems
}

func readStdIn() ([]string, error) {
	lines := []string{}
	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	if err := scanner.Err(); err != nil {
		return nil, fmt.Errorf("failed to read standard input: %w", err)
	}
	return lines, nil
}

func parseLines(lines []string) []CorrectOrNot {
	results := []CorrectOrNot{}
	for _, line := range lines {
		correctOrNot := CorrectOrNot{First: true, Second: true}
		line = strings.TrimSpace(line)
		elems := strings.Split(line, " ")
		if elems[0] == "n" {
			correctOrNot.First = false
		}
		if elems[1] == "n" {
			correctOrNot.Second = false
		}
	}
	return results
}

func main() {
	lines, err := readStdIn()
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("%v\n", lines)
	correctOrNots := parseLines(lines[1:])
	fmt.Printf("%v\n", correctOrNots)
	problems := new().Solve(correctOrNots)
	if len(problems) == 0 {
		fmt.Println("0")
	} else {
		fmt.Println(strconv.Itoa(len(problems)))
		for _, p := range problems {
			fmt.Println(strconv.Itoa(p))
		}
	}
}
